### ISSUE TITLE
ClientException thrown when tcp connection is closed

### ISSUE THREAD

**Issue by James-Ronayne** [microsoftgraph/msgraph-sdk-java#313]
Expected behavior
When a underlying connection okhttp is using is closed okhttp should reopen a new connection without throwing an exception.
Actual behavior
ClientException is thrown by the graph sdk.
Underlying connection okhttp is handling is closed causing a EOFException to be thrown. This eventually bubbles up to the sdk library and is caught and rethrown as a ClientException.
Steps to reproduce the behavior
Make a number of calls to the graph sdk consistantly over a period of 15-30 minutes and the error should be seen.
This looks to be caused by setting okBuilder.retryOnConnectionFailure(false); in CoreHttpProvider. I can't see why this was done but it happened when RetryHandler functionality was added from msgraph-sdk-java-core. Since RetryHandler only retries status code responses of 429 503 504 and doesn't actually do anything about exeptions this should be set to true.
There's no way to override this setting the only workaround I've found is to add retry logic into any code that makes a call with the sdk or inject a 'Connection: close' header into every request made making okhttp open a new connection everytime.
relevant info: square/okhttp/issues/1873
com.microsoft.graph.core.ClientException: Error during http request
com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:422) ~[microsoft-graph-1.6.0.jar!/:na]\tat com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:204) ~[microsoft-graph-1.6.0.jar!/:na]
 com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:184) ~[microsoft-graph-1.6.0.jar!/:na]
 com.microsoft.graph.http.BaseCollectionRequest.send(BaseCollectionRequest.java:89) ~[microsoft-graph-1.6.0.jar!/:na]
 com.microsoft.graph.requests.extensions.UserCollectionRequest.get(UserCollectionRequest.java:52) ~[microsoft-graph-1.6.0.jar!/:na]
.......
 org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61) [tomcat-embed-core-9.0.24.jar!/:9.0.24]
 java.lang.Thread.run(Thread.java:748) [na:1.8.0_222]
Caused by: java.io.IOException: unexpected end of stream on https://graph.microsoft.com/...
 okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:236) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.java:115) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:94) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:43) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:88) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:35) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31) ~[microsoft-graph-core-1.0.0.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:142) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:117) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:229) ~[okhttp-3.14.4.jar!/:na]
 okhttp3.RealCall.execute(RealCall.java:81) ~[okhttp-3.14.4.jar!/:na]
 com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:356) ~[microsoft-graph-1.6.0.jar!/:na]
... 105 common frames omitted
Caused by: java.io.EOFException: 
 not found: limit=0 content=…
 okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:240) ~[okio-1.17.2.jar!/:na]
 okhttp3.internal.http1.Http1ExchangeCodec.readHeaderLine(Http1ExchangeCodec.java:242) ~[okhttp-3.14.4.jar!/:na]\tat okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.java:213) ~[okhttp-3.14.4.jar!/:na]
\t... 134 common frames omitted
[AB#6038](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/6038)

**Comment by rnielsen** [comment]
Hi.  I'm also getting this error but for me it comes up as a `java.net.SocketException: Connection reset`.  Just thought I'd add a comment here in case anyone else is searching.

I've worked around it for now by using reflection to set the private coreHttpClient field in CoreHttpProvider to a OkHttpClient that has retryOnConnectionFailure(true) and it resolves the issue for me but a real fix would be good.

```
Error during http request: com.microsoft.graph.core.ClientException
com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:422)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:204)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:184)
	at com.microsoft.graph.http.BaseCollectionRequest.send(BaseCollectionRequest.java:89)
	at com.microsoft.graph.requests.extensions.SubscriptionCollectionRequest.get(SubscriptionCollectionRequest.java:52)
	at com.microsoft.graph.requests.extensions.ISubscriptionCollectionRequest$get.call(Unknown Source)
Caused by: java.net.SocketException: Connection reset
	at okio.Okio$1.write(Okio.java:79)
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:180)
```

**Comment by FelixJongleur42** [comment]
I am also getting some errors like this, they seem to be sporadic.
`Caused by: java.io.EOFException: \n not found: limit=0 content=…
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:240) ~[okio-1.17.2.jar:?]
	at okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:226) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:202) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:101) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:94) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.13.1.jar:?]
	at com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:35) ~[microsoft-graph-core-1.0.0.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.13.1.jar:?]
	at com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123) ~[microsoft-graph-core-1.0.0.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.13.1.jar:?]
	at com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140) ~[microsoft-graph-core-1.0.0.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.13.1.jar:?]
	at com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31) ~[microsoft-graph-core-1.0.0.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.13.1.jar:?]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.13.1.jar:?]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:264) ~[okhttp-3.13.1.jar:?]
	at okhttp3.RealCall.execute(RealCall.java:93) ~[okhttp-3.13.1.jar:?]
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:356) ~[microsoft-graph-1.7.1.jar:?]`

**Comment by guybmayor** [comment]
any update?

**Comment by xHatti** [comment]
Getting client errors too. This issue is so extreme that this client is almost not useable for me. The issue is totally random. Sometimes it works perfectly, sometimes it fails even after 15 retries..


Caused by: com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:422) ~[microsoft-graph-1.7.1.jar!/:na]
	at com.microsoft.graph.http.CoreHttpProvider.access$000(CoreHttpProvider.java:69) ~[microsoft-graph-1.7.1.jar!/:na]
	at com.microsoft.graph.http.CoreHttpProvider$1.run(CoreHttpProvider.java:155) ~[microsoft-graph-1.7.1.jar!/:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na]
	... 1 common frames omitted
Caused by: javax.net.ssl.SSLException: Connection reset by peer (Write failed)
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:127) ~[na:na]
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:326) ~[na:na]
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:269) ~[na:na]
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:264) ~[na:na]
	at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(SSLSocketImpl.java:980) ~[na:na]
	at okio.Okio$1.write(Okio.java:79) ~[okio-1.15.0.jar!/:na]
	at okio.AsyncTimeout$1.write(AsyncTimeout.java:180) ~[okio-1.15.0.jar!/:na]
	at okio.RealBufferedSink.flush(RealBufferedSink.java:224) ~[okio-1.15.0.jar!/:na]
	at okhttp3.internal.http1.Http1Codec.finishRequest(Http1Codec.java:166) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:84) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na]
	at com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:35) ~[microsoft-graph-core-1.0.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na]
	at com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123) ~[microsoft-graph-core-1.0.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na]
	at com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140) ~[microsoft-graph-core-1.0.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na]
	at com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31) ~[microsoft-graph-core-1.0.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254) ~[okhttp-3.12.1.jar!/:na]
	at okhttp3.RealCall.execute(RealCall.java:92) ~[okhttp-3.12.1.jar!/:na]
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:356) ~[microsoft-graph-1.7.1.jar!/:na]
	... 5 common frames omitted
	Suppressed: java.net.SocketException: Broken pipe (Write failed)
		at java.base/java.net.SocketOutputStream.socketWrite0(Native Method) ~[na:na]
		at java.base/java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:110) ~[na:na]
		at java.base/java.net.SocketOutputStream.write(SocketOutputStream.java:150) ~[na:na]
		at java.base/sun.security.ssl.SSLSocketOutputRecord.encodeAlert(SSLSocketOutputRecord.java:81) ~[na:na]
		at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:357) ~[na:na]
		... 40 common frames omitted
Caused by: java.net.SocketException: Connection reset by peer (Write failed)
	at java.base/java.net.SocketOutputStream.socketWrite0(Native Method) ~[na:na]
	at java.base/java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:110) ~[na:na]
	at java.base/java.net.SocketOutputStream.write(SocketOutputStream.java:150) ~[na:na]
	at java.base/sun.security.ssl.SSLSocketOutputRecord.deliver(SSLSocketOutputRecord.java:319) ~[na:na]
	at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(SSLSocketImpl.java:975) ~[na:na]

**Comment by rukletsov** [comment]
We're observing the same issue for months now. I'm going to implement an infinite retry loop based on the suggestions above. It would be great to get an estimate on the fix and a suggestion for a workaround. @bilalgce maybe?

**Comment by kekolab** [comment]
Similar problems here, but the error disappears using good old java 8.152. The error seems present only from java 9 and upwards and seems related to the fact that okhttp from this java version negotiates by default http2 instead of http1.1. Indeed with the debugger on, I intercepted the creation of the okhttp client in CoreHttpTansport and limited the negotiable http versions to only http1.1 and the errors disappeared on java9 too.

Still looking for a way to have it working automatically.

**Comment by FelixJongleur42** [comment]
> limited the negotiable http versions to only http1.1 and the errors disappeared on java9 too.

that is some great insight!

Now can someone from Microsoft go ahead and have a look at this???

**Comment by rukletsov** [comment]
@MIchaelMainer any chance this ticket gets some attention?

**Comment by kekolab** [comment]
Hello everyone interested,
I have been trying for some days to use the Apache Commons HttpClient as an IHttpProvider. I also managed to use it successfully in all the stateless requests. Problems arise when you try to implement the following [send](https://javadoc.io/static/com.microsoft.graph/microsoft-graph/1.5.0/com/microsoft/graph/http/IHttpProvider.html#send-com.microsoft.graph.http.IHttpRequest-java.lang.Class-BodyType-com.microsoft.graph.http.IStatefulResponseHandler-) function of IHttpProvider
````
<Result,BodyType,DeserializeType> Result send(IHttpRequest request, java.lang.Class<Result> resultClass, BodyType serializable, IStatefulResponseHandler<Result,DeserializeType> handler) throws ClientException
````
where you are supposed to send the request to the remote server, receive the reponse and have it handled by the handler. The problem is that the [handling function](https://javadoc.io/static/com.microsoft.graph/microsoft-graph/1.5.0/com/microsoft/graph/http/IStatefulResponseHandler.html#configConnection-okhttp3.Response-) signature is: 
````
ResultType generateResult(IHttpRequest request, okhttp3.Response response, ISerializer serializer, ILogger logger) throws java.lang.Exception
````
i.e. the second parameter is not an interface to implement, but the class the OKHttpClient returns after receiving the response from the remote server, so Apache Commons Client's HttpResponse cannot be either used or adapted easily.

My **ugly** solution to have the APIs, at the end is by using reflection to impose to the CoreHttpProvider an OKHttpClient which can negotiate only HTTP/1.1. Hereby the code:
````
this.graphServiceClient = GraphServiceClient.builder().authenticationProvider(authenticationProvider).buildClient()
try {
    /* 
     * Reflection to access the httpProvider
     */
    Field httpProviderField = BaseClient.class.getDeclaredField("httpProvider");
    httpProviderField.setAccessible(true);
    CoreHttpProvider httpProvider = (CoreHttpProvider) httpProviderField.get(graphServiceClient);
			
    /*
     * Reflection to access the coreHttpProvider's coreHttpClient
     */
    Field coreHttpClientField = CoreHttpProvider.class.getDeclaredField("corehttpClient");
    coreHttpClientField.setAccessible(true);
    OkHttpClient coreHttpClient = (OkHttpClient) coreHttpClientField.get(httpProvider);
			
    /*
     * Building an OKHttpClient without HTTP/2
    */
    if(coreHttpClient == null) {
        /*
         * Reflection to get the coreHttpProvider's connectionConfig
         */
        Field connectionConfigField = CoreHttpProvider.class.getDeclaredField("connectionConfig");
        connectionConfigField.setAccessible(true);
        IConnectionConfig connectionConfig = (IConnectionConfig) connectionConfigField.get(httpProvider);
				
        /*
         * Initing connection config to default if null
         */
        if(connectionConfig == null)
            connectionConfig = new DefaultConnectionConfig();
				
        /*
         * Reflection to set the coreHttpProviver's coreHttpProvider to an HTTP1.1 only OKHttpClient
         * The code is copypasted from CoreHttpProvider. I have just added the line 
         * .protocols(Arrays.asList(Protocol.HTTP_1_1)) so that the OKHttpClient cannot negotiate
         * HTTP/2
         */
        coreHttpClientField.set(httpProvider, HttpClients
            .createDefault(request -> request)
            .newBuilder()
            .protocols(Arrays.asList(Protocol.HTTP_1_1))
            .connectTimeout(connectionConfig.getConnectTimeout(), TimeUnit.MILLISECONDS)
            .readTimeout(connectionConfig.getReadTimeout(), TimeUnit.MILLISECONDS)
            .followRedirects(false)
            .retryOnConnectionFailure(false)
            .build());
    }
} catch (Exception e) {
    e.printStackTrace();
}
````
As I said, just like everything done with reflection, the solution is extremely **ugly** neither it can be considered **definitive** as in case you have a security manager, it won't work.
Though, if you do not have a security manager, the OKHttpClient will work exactly like on Java <9, i.e. **without any error and with a higher transfer speed** (in my case, about 3Mbps instead of 1Mbps). 

I do not know if the error is caused by some bug in the handling of HTTP/2 in OKHttpClient or in some misconfigurations in Microsoft servers. But at least now everything works also on Java 9+

**Comment by samdohrey** [comment]
Hi All,

Same issue here. Had a debug environment running for 5 hours, it polls an email box twice every 2 minutes and the error appeared

```
com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:422)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:204)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:184)
	at com.microsoft.graph.http.BaseCollectionRequest.send(BaseCollectionRequest.java:89)
	at com.microsoft.graph.requests.extensions.MessageCollectionRequest.get(MessageCollectionRequest.java:52)

```

Further down the error stack:

```
Caused by: javax.net.ssl.SSLException: Connection reset by peer: socket write error
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:127)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:320)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:263)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:258)
	at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(SSLSocketImpl.java:1016)
	at okio.OutputStreamSink.write(JvmOkio.kt:53)
	at okio.AsyncTimeout$sink$1.write(AsyncTimeout.kt:103)
	at okio.RealBufferedSink.flush(RealBufferedSink.kt:247)
	at okhttp3.internal.http1.Http1ExchangeCodec.finishRequest(Http1ExchangeCodec.kt:155)
```

**Comment by zengin** [comment]
We received reports of similar issues recently, and we should look into ways of incorporating a retry mechanism to work around these failures.

**Comment by lespea** [comment]
Just another point to add... I'm currently migrating from EWS and even there I constantly get `Connection reset by peer: socket write error`, so I don't think it's limited to just Graph.

**Comment by ravisamani** [comment]
@lespea  You are right.

we're continuously getting such errors for both of below singleton HTTP clients -

Apache HttpClient
IGraphServiceClient

**Comment by juanmarques** [comment]
Hello , I just started to use the sdk , but sometimes I get this error bellow and I think it's related.

`com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:422)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:204)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:184)
	at com.microsoft.graph.http.BaseStreamRequest.send(BaseStreamRequest.java:85)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:497)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:487)
	at java.base/java.util.stream.ForEachOps$ForEachOrderedTask.onCompletion(ForEachOps.java:494)
	at java.base/java.util.concurrent.CountedCompleter.tryComplete(CountedCompleter.java:592)
	at java.base/java.util.stream.ForEachOps$ForEachOrderedTask.doCompute(ForEachOps.java:482)
	at java.base/java.util.stream.ForEachOps$ForEachOrderedTask.compute(ForEachOps.java:399)
	at java.base/java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:746)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.helpCC(ForkJoinPool.java:1115)
	at java.base/java.util.concurrent.ForkJoinPool.externalHelpComplete(ForkJoinPool.java:1957)
	at java.base/java.util.concurrent.ForkJoinTask.tryExternalHelp(ForkJoinTask.java:378)
	at java.base/java.util.concurrent.ForkJoinTask.externalAwaitDone(ForkJoinTask.java:323)
	at java.base/java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:412)
	at java.base/java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:736)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:157)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:238)
	at java.base/java.util.stream.ReferencePipeline.forEachOrdered(ReferencePipeline.java:502)
	at java.base/java.util.stream.ReferencePipeline$Head.forEachOrdered(ReferencePipeline.java:671)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1540)
	at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: okhttp3.internal.http2.ConnectionShutdownException
	at okhttp3.internal.http2.Http2Connection.newStream(Http2Connection.java:248)
	at okhttp3.internal.http2.Http2Connection.newStream(Http2Connection.java:231)
	at okhttp3.internal.http2.Http2Codec.writeRequestHeaders(Http2Codec.java:117)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:50)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:37)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254)
	at okhttp3.RealCall.execute(RealCall.java:92)
	at com.microsoft.gr`

**Comment by axk7812** [comment]
This would be awesome if could be addressed! Currently, I'm doing an infinite retry loop as well but it's still concerning.

**Comment by baywet** [comment]
Hi everyone,
Thanks for your patience on the matter and for the extended feedback.
I'm still looking into this but here is what I found so far.
The [change](https://github.com/microsoftgraph/msgraph-sdk-java/commit/7253bd90e70de6c6d9216d7a69a214905cf74a45) was added by #254.
After chatting internally with the engineer who made the change back then, as far as he remembers this change was introduced to avoid redundancy with the [retry handler](https://github.com/microsoftgraph/msgraph-sdk-java-core/blob/dev/src/main/java/com/microsoft/graph/httpcore/RetryHandler.java) and respect the [following spec](https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/middleware/RetryHandler.md).
Now, [reading the docs](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-ok-http-client/-builder/retry-on-connection-failure/) is doesn't sound like this setting would be conflicting in any shape or form with the spec. I have some testing to do though.

**Comment by baywet** [comment]
Hi everyone,
After some extended testing it looks like this setting and the handler are not overlapping.
I authored PR #473 which fixes this issue. Merging the PR will close this issue.
I also authored https://github.com/microsoftgraph/msgraph-sdk-java-core/pull/74 which is related.

**Comment by ravisamani** [comment]
After upgrading core to 1.2.0 and SDK to 2.1.0 - we continue to get - 

`Caused by: java.net.SocketException: Connection reset by peer (Write failed)
	at java.base/java.net.SocketOutputStream.socketWrite0(Native Method)
	at java.base/java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:110)
	at java.base/java.net.SocketOutputStream.write(SocketOutputStream.java:150)
	at java.base/sun.security.ssl.SSLSocketOutputRecord.deliver(SSLSocketOutputRecord.java:319)
	at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(SSLSocketImpl.java:983)`

@baywet 

So root cause for this is in retry handler you mentioned above?

**Comment by baywet** [comment]
@ravisamani that's normal. The pull request hasn't been merged yet. It should go out with 2.2.0.

Without knowing too many of the details. I know the service can drop connections sometimes when the instance you're connected to gets too hot. It does that to force the client to reconnect and land on another instance. That's the main explanation I can see for connections being dropped assuming this is not coming from a network issue and/or an issue with the client OS.

I went through that thread, it was really interesting. I'm a bit reluctant to disable http2 as if we do that, it'll require the library to be upgraded in the future when the service supports http/2. Right now, whenever the service supports it, all the client apps already out there should naturally transition without any modification required and it should represent a performance gain "for free".

**Comment by baywet** [comment]
update: after doing extended testing for #481 it appears a number of services reset the connection when initiated with HTTP/2 from the client. I have added the protocol restriction to #473. My guess is we'll have to do some extended testing the day all services support HTTP/2 and we decide to enable it in the client library. Unfortunately that'll require a dependency upgrade and a distribution to get the benefits.

**Comment by kekolab** [comment]
@baywet That's the way to go. See my comment above. Following the response to my question on stackoverflow (which you have actually referenced in the code too) I made the same modification (using reflection in my app, not modifying the library) and my app is still working flawlessly after several months.

Glad to know that updating the dependency I will be able to get rid of the quick dirty and ugly approach I implemented in my code

**Comment by ravisamani** [comment]
@baywet  Hi Vincent,

any clue when is 2.2.0 releasing? as **2.2.0** shows 18 Sept milestone here https://github.com/microsoftgraph/msgraph-sdk-java/milestone/5

**Comment by baywet** [comment]
The original date was the 25th. I had updated it early last week thinking we would be able to release it earlier but the last issues were harder to tackle than I thought.
We're still targeting a release sometime this week. I'll update the release date later today.

**Comment by ravisamani** [comment]
Hi @baywet,

After upgrading to 2.2.0 -  **connection reset** errors are resolved.  

But now, we started facing **SocketException: Socket closed** error repeatedly for multiple APIs.

Any input?


Full trace:
`com.microsoft.graph.core.ClientException: Error during http request
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:458)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:207)
	at com.microsoft.graph.http.CoreHttpProvider.send(CoreHttpProvider.java:187)
	at com.microsoft.graph.http.BaseRequest.send(BaseRequest.java:345)
	at com.microsoft.graph.http.CustomRequest.get(CustomRequest.java:43)
	at com.optymyze.officeintegration.alternateteams.GraphApiBatchService.getOrRetryGraphApiRequest(GraphApiBatchService.java:130)
	...	15 more
Caused by: java.net.SocketTimeoutException: timeout
	at okio.Okio$4.newTimeoutException(Okio.java:232)
	at okio.AsyncTimeout.exit(AsyncTimeout.java:285)
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:241)
	at okio.RealBufferedSource.indexOf(RealBufferedSource.java:354)
	at okio.RealBufferedSource.readUtf8LineStrict(RealBufferedSource.java:226)
	at okhttp3.internal.http1.Http1Codec.readHeaderLine(Http1Codec.java:215)
	at okhttp3.internal.http1.Http1Codec.readResponseHeaders(Http1Codec.java:189)
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:88)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:37)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254)
	at okhttp3.RealCall.execute(RealCall.java:92)
	at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:385)
	... 45 more
Caused by: javax.net.ssl.SSLException: Socket closed
	at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:127)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:320)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:263)
	at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:258)
	at java.base/sun.security.ssl.SSLSocketImpl.handleException(SSLSocketImpl.java:1314)
	at java.base/sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:839)
	at okio.Okio$2.read(Okio.java:140)
	at okio.AsyncTimeout$2.read(AsyncTimeout.java:237)
	... 77 more
Caused by: java.net.SocketException: Socket closed
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:183)
	at java.base/java.net.SocketInputStream.read(SocketInputStream.java:140)
	at java.base/sun.security.ssl.SSLSocketInputRecord.read(SSLSocketInputRecord.java:448)
	at java.base/sun.security.ssl.SSLSocketInputRecord.bytesInCompletePacket(SSLSocketInputRecord.java:68)
	at java.base/sun.security.ssl.SSLSocketImpl.readApplicationRecord(SSLSocketImpl.java:1104)
	at java.base/sun.security.ssl.SSLSocketImpl$AppInputStream.read(SSLSocketImpl.java:823)
	... 79 more`

**Comment by baywet** [comment]
@ravisamani can you run gradlew dependencies and paste the output here please?
What kind of application are you building? (android, backend, API, something else?)

**Comment by ravisamani** [comment]
@baywet Backend Application.
 
you need dependency tree with to see library versions?

**Comment by baywet** [comment]
yes, I'm trying to understand which version of core and okhttp are in use

**Comment by shahamit** [comment]
@baywet - The library versions are 
core - 1.0.5
okhttp - 3.12.1

Other details
"com.microsoft.graph:microsoft-graph:2.3.0",
"com.squareup.okio:okio:1.15.0"

P.S - @ravisamani has been unwell. Hence I am posting being part of his team.

**Comment by baywet** [comment]
thanks for the follow up @shahamit .
How often are you instantiating the GraphServiceClient as part of your application? Is your application running on windows or linux?

**Comment by shahamit** [comment]
We are running on Linux. We instantiate the client just once. I will still
review the code once and let you know if otherwise

**Comment by shahamit** [comment]
@baywet - I reviewed the code and I would like to confirm that we are creating the service client just once in our application.

I came across [this ](https://microsoftgraph.github.io/msgraph-sdk-design/GraphClientFactory.html)article that states that authorization tokens should not be set as default (not sure what this means?) request headers. We do set the token as a request header like the [sample code snippets](https://github.com/microsoftgraph/msgraph-training-java/blob/master/demo/graphtutorial/src/main/java/graphtutorial/SimpleAuthProvider.java) also suggest. Also we have TLS 1.2 configured at our end though we don't pass it in the request header. Do we need to?

1. Could you please let us know if there are any changes we need to make in our code?
2. We are seeing many "socket closed" errors in our testing and production environments and hence are looking forward for some resolution quickly. There doesn't seem to be any pattern that we could find with the errors. Do you want to create a new issue or reopen this in order to prioritize this issue?

Thanks.

**Comment by baywet** [comment]
# TLS
We're using the OkHttp [which uses all the the available TLS versions](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-connection-spec/tls-versions/) if no setting is set (and the graph SDK doesn't set any version).
With TLS, a client will use the [uppermost version it and the server have in common by default](https://blog.cloudflare.com/why-tls-1-3-isnt-in-browsers-yet/#:~:text=TLS%20version%20negotiation%20is%20very%20simple.%20The%20client,Versions%20in%20TLS%20are%20represented%20as%20two-byte%20values.).
Microsoft Graph [supports up to TLS 1.2](https://www.ssllabs.com/ssltest/analyze.html?d=graph.microsoft.com&s=20.190.132.118).

I think it's safe to assume that if you haven't set anything for the OkHttpClient configuration, the JVM configuration and if your OS supports TLS 1.2, all communications are happening with TLS 1.2.
**One thing you could do on that side is double check that your OS supports TLS 1.2 and that the JVM is up to date + not configured to disable TLS 1.2.**

# Auth
The guidelines you pointed at are guidelines provided by our program manager for the SDK development team. We share it publicly so people know how things are designed and implemented and can contribution to design decisions, but those directives are more "for me" than "for you". That's unless you implemented your own authentication provider instead of relying on the ones we provide?
**The sample should check for the hostname before attaching the token as the same client is used for things like uploads sessions that go to a different hostname.**

# Socket closed exception
From what I can see from the stacktrace you share it looks like the server is [forcibly closing the connection](https://github.com/square/okio/issues/702) as it originates from [executing the request](https://github.com/microsoftgraph/msgraph-sdk-java/blob/2.2.0/src/main/java/com/microsoft/graph/http/CoreHttpProvider.java#L385). This is something the service does to protect itself sometimes whenever a specific application has too many parallel requests (that's in your control) or if a specific service instance/server is getting overwhelmed (this is not entirely in your control as it depends on other apps using the service).
Now, I'm not sure whether the okhttp library is retrying, which [it should do by default](https://medium.com/inloopx/okhttp-is-quietly-retrying-requests-is-your-api-ready-19489ef35ace). But the presence of RetryAndFollowUpInterceptor in the callstack leads me to think it is retry. (maybe too aggressively?)
Our [retry handler](https://github.com/microsoftgraph/msgraph-sdk-java-core/blob/dev/src/main/java/com/microsoft/graph/httpcore/RetryHandler.java) is only designed to retry on specific error codes to avoid mixing concerns.

With all these elements in mind, here are my follow up questions for you:
- Are you doing any form of multithreading in your application?
- How distributed is your application?
- Can you share the code of the request? (as you're writing a custom request)

# Creating a new issue
Yes I think it'd be interesting to create a new issue as this is a new problem that has been revealed by fixing the one expose in this original issue. It'll get better visibility.

**Comment by shahamit** [comment]
Appreciate the detailed analysis on the issue.

1. We will verify the TLS settings on the OS & JVM
2. I didn't quiet follow how do we validate the hostname

> The sample should check for the hostname before attaching the token as the same client is used for things like uploads sessions that go to a different hostname.

Would the IHttpRequest object have the hostname? Even if it does whom do we compare them with?

3. Disabling the default retry from okhttp client seems like an immediate thing that we can try out. We will let you know if this stops any exceptions from being reported.

Answering your questions

> Are you doing any form of multithreading in your application?

No, we don't do multithreading when making the graph api request but we do have multiple instances of our application running in parallel that may be causing multiple api requests being made at the same time. We have done our best to make sure that the api calls are not being made at the same time but they may clash in the worst case scenario. Let us analyze this further if disabling the retries from okhttp doesn't help.

> How distributed is your application?

In most of the cases we have one app server (tomcat) making the graph api calls but the application is deployed individually for each of our clients so there are multiple instances of the application running at the same time.

> Can you share the code of the request? (as you're writing a custom request)

Sure, below are some samples of the api requests being made by our code
Sample 1
`JsonObject jsonObject = factory.graphClient(USER).me()
                .buildRequest(newArrayList(new QueryOption(SELECT, MS_USER_ATTRIBUTES_TO_FETCH))).get().getRawObject();`
Sample 2
`JsonObject jsonObject = factory.graphClient(APPLICATION)
                .customRequest(format(O365_GRAPH_API_GET_JOINED_TEAMS_FOR_USER, userId))
                .buildRequest().get();`

**Comment by shahamit** [comment]
@baywet - As stated on [this ](https://medium.com/inloopx/okhttp-is-quietly-retrying-requests-is-your-api-ready-19489ef35ace)link, we could disable retries on okhttp client but since our code does not interact with OkHttpClient directly, could you help us how do we disable the retries through the graph library?

**Comment by baywet** [comment]
Checking for the hostname: have a look at [this example](https://github.com/microsoftgraph/msgraph-sdk-java/blob/dev/src/test/java/com/microsoft/graph/functional/TestBaseAuthenticationProvider.java)

Disabling retries: sure, you'll run into the initial issue this thread was opened for though. Have a look at [this documentation](https://docs.microsoft.com/en-us/graph/sdks/customize-client?tabs=java) once you have the okhttpclient you can setup pretty much anything you want before passing it back to the graph service client builder.

Multithreads/multi-instances: if you're distributed on multiple "machines" (i.e. calling from different public IPs) it should be fine, the service doesn't like the same IP making too many parallel connections and has a tendency to drop them when that occurs.  There are other protection mechanisms in place if an application calls Microsoft Graph too often (even from multiple IPs) but I'm guessing you'd have run into throttling issues first.

Thanks for sharing the samples. Out of curiosity, why don't you do this instead for the first sample? (using high level objects instead of raw json objects, using the provided fluent API instead of a custom request builder)
```Java
final User myUser = factory.graphClient().me().buildRequest().select(MS_USER_ATTRIBUTES_TO_FETCH).get()
```

**Comment by shahamit** [comment]
> Checking for the hostname: have a look at this example

I see. I know understand how to add the hostname check but I don't follow the underlying reason behind doing it. You mentioned that the same client could be used for upload sessions - are you referring to file uploads to sharepoint? If so, what is the correlation and why should we skip adding the authorization header for such requests? Won't the request fail with 401?

> Have a look at this documentation once you have the okhttpclient you can setup pretty much anything you want before passing it back to the graph service client builder.

Thanks for sharing the code. We will try it and follow up.

> There are other protection mechanisms in place if an application calls Microsoft Graph too often (even from multiple IPs) but I'm guessing you'd have run into throttling issues first.

Yes we have seen API rate limiting errors (HTTP 429) multiple times in the past and hence we are even caching a lot of data. 
Will follow up on other points tomorrow.

**Comment by shahamit** [comment]
> Out of curiosity, why don't you do this instead for the first sample?

Yes, sure. The code sample that I shared was most probably based on an older version of the graph sdk

> if you're distributed on multiple "machines" (i.e. calling from different public IPs) it should be fine, the service doesn't like the same IP making too many parallel connections and has a tendency to drop them when that occurs. 

I double checked with our deployment team. All our application servers are behind the firewall hence the graph server would see only one public IP address for all requests and this is kind off a standard deployment model used by everyone - the app servers are always on a private IP address.  Are you sure that the graph api service uses just the IP address for API rate limiting or are there other attributes to detect client request uniqueness?

Kindly let me know your thoughts on this and previous comments too.

**Comment by baywet** [comment]
**hostname**: if you use [upload sessions](https://docs.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0) the hostname will be something like `https://sn3302.up.1drv.com/up/fe6987415ace7X4e1eF866337` (for OneDrive, each service has their own upload URLs). The access token you get is only valid for a specific audience (graph.microsoft.com) and not valid for any other domain. Sending it along with upload requests will cause an access denied. The reason why you don't need an access token for those URLs is because they are either short-lived or they contain the access token as a query string parameter. (one valid for that audience)

**public IP**: Microsoft Graph has multiple checks in place to protect itself from malicious actors or applications that are too intensive. Without being able to share the specifics publicly (number of connections, etc...), if the service sees too many connections originating from the same IP, it will start dropping connections. Are all your customer instances behind the same firewall?

What did your investigations disabling the retry resulted in?

**Comment by shahamit** [comment]
@baywet - I have logged a new issue for this. Let's continue the discussion on that thread.

**Comment by Ahmid** [comment]
> Getting client errors too. This issue is so extreme that this client is almost not useable for me. The issue is totally random. Sometimes it works perfectly, sometimes it fails even after 15 retries..
> 
> Caused by: com.microsoft.graph.core.ClientException: Error during http request at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:422) ~[microsoft-graph-1.7.1.jar!/:na] at com.microsoft.graph.http.CoreHttpProvider.access$000(CoreHttpProvider.java:69) ~[microsoft-graph-1.7.1.jar!/:na] at com.microsoft.graph.http.CoreHttpProvider$1.run(CoreHttpProvider.java:155) ~[microsoft-graph-1.7.1.jar!/:na] at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[na:na] at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[na:na] ... 1 common frames omitted Caused by: javax.net.ssl.SSLException: Connection reset by peer (Write failed) at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:127) ~[na:na] at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:326) ~[na:na] at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:269) ~[na:na] at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:264) ~[na:na] at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(SSLSocketImpl.java:980) ~[na:na] at okio.Okio$1.write(Okio.java:79) ~[okio-1.15.0.jar!/:na] at okio.AsyncTimeout$1.write(AsyncTimeout.java:180) ~[okio-1.15.0.jar!/:na] at okio.RealBufferedSink.flush(RealBufferedSink.java:224) ~[okio-1.15.0.jar!/:na] at okhttp3.internal.http1.Http1Codec.finishRequest(Http1Codec.java:166) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:84) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na] at com.microsoft.graph.httpcore.TelemetryHandler.intercept(TelemetryHandler.java:35) ~[microsoft-graph-core-1.0.0.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na] at com.microsoft.graph.httpcore.RedirectHandler.intercept(RedirectHandler.java:123) ~[microsoft-graph-core-1.0.0.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na] at com.microsoft.graph.httpcore.RetryHandler.intercept(RetryHandler.java:140) ~[microsoft-graph-core-1.0.0.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na] at com.microsoft.graph.httpcore.AuthenticationHandler.intercept(AuthenticationHandler.java:31) ~[microsoft-graph-core-1.0.0.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147) ~[okhttp-3.12.1.jar!/:na] at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121) ~[okhttp-3.12.1.jar!/:na] at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:254) ~[okhttp-3.12.1.jar!/:na] at okhttp3.RealCall.execute(RealCall.java:92) ~[okhttp-3.12.1.jar!/:na] at com.microsoft.graph.http.CoreHttpProvider.sendRequestInternal(CoreHttpProvider.java:356) ~[microsoft-graph-1.7.1.jar!/:na] ... 5 common frames omitted Suppressed: java.net.SocketException: Broken pipe (Write failed) at java.base/java.net.SocketOutputStream.socketWrite0(Native Method) ~[na:na] at java.base/java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:110) ~[na:na] at java.base/java.net.SocketOutputStream.write(SocketOutputStream.java:150) ~[na:na] at java.base/sun.security.ssl.SSLSocketOutputRecord.encodeAlert(SSLSocketOutputRecord.java:81) ~[na:na] at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:357) ~[na:na] ... 40 common frames omitted Caused by: java.net.SocketException: Connection reset by peer (Write failed) at java.base/java.net.SocketOutputStream.socketWrite0(Native Method) ~[na:na] at java.base/java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:110) ~[na:na] at java.base/java.net.SocketOutputStream.write(SocketOutputStream.java:150) ~[na:na] at java.base/sun.security.ssl.SSLSocketOutputRecord.deliver(SSLSocketOutputRecord.java:319) ~[na:na] at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(SSLSocketImpl.java:975) ~[na:na]

Hi @xHatti , We are facing the exact same issue. How dod you solve it?

**Comment by ckarawani** [comment]
@Ahmid last we saw this was resolved by https://github.com/microsoftgraph/msgraph-sdk-java-core/pull/74

**Comment by xHatti** [comment]
@Ahmid used newer version of OKHttpClient (4.10.0) currently
And gave the Graphservice a own build http client that looks like that

`

	Builder clientBuilder = baseHttpClient.newBuilder()//
		.addInterceptor(new TelemetryHandler()) //
		.followRedirects(false) //
		.followSslRedirects(false) //
		.protocols(Arrays.asList(Protocol.HTTP_1_1)) //https://stackoverflow.com/questions/62031298/sockettimeout-on-java-11-but-not-on-java-8
		.addInterceptor(new AuthenticationHandler(authProvider)) //
		.addInterceptor(new RetryHandler(new RetryOptions(RetryOptions.DEFAULT_SHOULD_RETRY, 3, RetryOptions.DEFAULT_DELAY))) //
		.addInterceptor(new RedirectHandler());
`
